soh/soh/Enhancements/randomizer/3drando/item_pool.cpp

```diff
@@ -1164,118 +768,117 @@ void GenerateItemPool() {
+            switch (ctx->GetOption(RSK_ITEM_POOL).Get()) {
+                case RO_ITEM_POOL_PLENTIFUL:
+                case RO_ITEM_POOL_MINIMAL:
+                    AddFixedItemToPool(RG_HEART_CONTAINER, maxHearts - startingHearts, false);
+                    break;
+                case RO_ITEM_POOL_BALANCED: {
+                    int heartsToPlace = maxHearts - startingHearts;
+                    int halfHearts = maxHearts >> 2;
+                    AddFixedItemToPool(RG_HEART_CONTAINER, heartsToPlace - halfHearts, false);
+                    AddFixedItemToPool(RG_PIECE_OF_HEART, halfHearts * 4, false);
+                    break;
+                }
+                case RO_ITEM_POOL_SCARCE:
+                    AddFixedItemToPool(RG_PIECE_OF_HEART, (maxHearts - startingHearts) * 4, false);
+                    break;
             }
         }
     }
 
-    if (ctx->GetOption(RSK_BOSS_KEYSANITY).Is(RO_DUNGEON_ITEM_LOC_VANILLA)) {
-        PlaceVanillaBossKeys();
-    } else {
-        AddItemToMainPool(RG_FOREST_TEMPLE_BOSS_KEY);
-        AddItemToMainPool(RG_FIRE_TEMPLE_BOSS_KEY);
-        AddItemToMainPool(RG_WATER_TEMPLE_BOSS_KEY);
-        AddItemToMainPool(RG_SPIRIT_TEMPLE_BOSS_KEY);
-        AddItemToMainPool(RG_SHADOW_TEMPLE_BOSS_KEY);
-    }
+    std::erase(junkPool, RG_NONE);
+    std::erase(itemPool, RG_NONE);
+    std::erase(lesserPool, RG_NONE);
+    std::erase(plentifulPool, RG_NONE);
 
-    if (!ctx->GetOption(RSK_TRIFORCE_HUNT)) { // Don't add GBK to the pool at all for Triforce Hunt.
-        if (ctx->GetOption(RSK_GANONS_BOSS_KEY).Is(RO_GANON_BOSS_KEY_KAK_TOKENS)) {
-            ctx->PlaceItemInLocation(RC_KAK_100_GOLD_SKULLTULA_REWARD, RG_GANONS_CASTLE_BOSS_KEY);
-        } else if (ctx->GetOption(RSK_GANONS_BOSS_KEY).Get() >= RO_GANON_BOSS_KEY_LACS_VANILLA) {
-            ctx->PlaceItemInLocation(RC_TOT_LIGHT_ARROWS_CUTSCENE, RG_GANONS_CASTLE_BOSS_KEY);
-        } else if (ctx->GetOption(RSK_GANONS_BOSS_KEY).Is(RO_GANON_BOSS_KEY_VANILLA)) {
-            ctx->PlaceItemInLocation(RC_GANONS_TOWER_BOSS_KEY_CHEST, RG_GANONS_CASTLE_BOSS_KEY);
+    size_t locCount = ctx->CountEmptyLocations(false);
+    assert(itemPool.size() <= locCount);
+    int iceTrapstoAdd = 0;
+    if (itemPool.size() + plentifulPool.size() < locCount) {
+        itemPool.insert(itemPool.end(), plentifulPool.begin(), plentifulPool.end());
+        // Fixed Ice Traps
+        if (ctx->GetOption(RSK_BASE_ICE_TRAPS)) {
+            iceTrapstoAdd++;
+            if (ctx->GetDungeon(Rando::GERUDO_TRAINING_GROUND)->IsVanilla()) {
+                iceTrapstoAdd++;
+            }
+            if (ctx->GetDungeon(Rando::GANONS_CASTLE)->IsVanilla()) {
+                iceTrapstoAdd += 4;
+            }
+        }
+        iceTrapstoAdd += ctx->GetOption(RSK_ICE_TRAP_COUNT).Get();
+        AddItemToMainPool(RG_ICE_TRAP,
+                          itemPool.size() + iceTrapstoAdd < locCount ? iceTrapstoAdd : locCount - itemPool.size());
```

Comment:
```suggestion
        AddItemToMainPool(RG_ICE_TRAP,
                          itemPool.size() + iceTrapstoAdd < locCount ? iceTrapstoAdd : locCount - itemPool.size(), false);
```
To avoid ice traps disguising as ice traps.

Reply:
this was actually using an old function instead of the new one that automatically adds to the ice trap pool, but removing that old function is probably better anyway, so I'll convert it over.